### ISSUE TITLE
Lots of issues with arrays

### ISSUE THREAD

**Issue by drmeister** [drmeister/clasp#125]
```
09:15 <stassats> ok, it won't really work
09:15 ⇐ sdemarre quit (~serge@79.191-64-87.adsl-dyn.isp.belgacom.be) Ping timeout: 240 seconds
09:15 <stassats> File(include/clasp/core/lispVector.h) lineNumber(63): Subclass[COMMON-LISP:BASE-STRING] must implement method[swapElements]
09:16 <stassats> drmeister: (nreverse "abc") doesn't work
09:17 <stassats> neither does (reverse "abc") but with a different reason
09:17 <stassats> (nreverse #(a b c)) produces #<vector 0x7fb989c29e49> 
09:18 <stassats> drmeister: so, i'd say that would be a prime target for fixing
09:19 <stassats> (aref "abc" -1) => #\NUL
09:19 <stassats> that's not right
09:19 ⇐ shka quit (~herr@109.231.38.106) Remote host closed the connection
09:20 <stassats> (aref "abc" 100) => #\E
09:20 <stassats> totally not right
09:20 <stassats>  (svref "abc" 0) => #\a, you guessed it, not right
09:22 <stassats> nreverse also doesn't work on #*110
09:23 <stassats> (concatenate 'string '(1 2 3)) says "expected class_id 68 symbol was not defined"
09:24 <stassats> (coerce "abc" 'bit-vector) produces #<vector 0x7fcff5da59f9>
09:24 <drmeister> I need coffee
09:25 <stassats> but even the right thing (coerce '(1 0 1) 'bit-vector) produces #<vector 0x7fcff83c75a9>
09:25 <stassats> (sbit "abc" 0) => #\a
09:26 <stassats> (subseq #*101 0 1) doesn't work
09:27 <stassats> (subseq #(a b c) 0 1) gives #<vector 0x7fcff830d579>
09:27 <stassats> (equalp #(a b c) #(a b c)) says it's not implemented
09:28 <stassats> (equal #*111 #*111) => T , but (equalp #*111 #*111) says it's not implemented
09:30 <stassats> (equalp 1 1.0) returns NIL, while it should be T, since (= 1 1.0) => T
09:30 <stassats> are you keeping up?
09:31 <stassats> (eql -0.0 0.0) says T, but it should be NIL
09:33 <stassats> huh, (time (eval '(= 3 5))) takes longer than (time (eval '(list 3 5)))
09:34 <stassats> but only in slime?
09:36 <stassats> ok, not only in slime, only in cclasp, in bclasp they are the same
09:40 <stassats> (time (eval '(funcall '= 3 5))) is faster than (time (eval '(= 3 5)))
09:41 <stassats> are you expanding compiler macros in the evaluator?
09:46 <stassats> so, your EVAL does full compilation in this case, and compiling inlined functions is slow
09:47 <stassats> drmeister: why aren't you using cclasp-eval for eval? it will be much faster for simple cases like this
09:48 <stassats> but anyway, (time (compile nil '(lambda (x y) (list x y)))) =>  0.097 secs
09:48 <stassats> (time (compile nil '(lambda (x y) (= x y)))) : 0.206 secs
09:54 ⇐ Guest70392 quit (~vibs29@46.233.112.199) Quit: This computer has gone to sleep
09:55 <stassats> M-. stopped working
```

**Comment by Shinmera** [comment]
Can confirm most on b05884f.
```
> (nreverse "abc")

"cba"
> (reverse "abc")

"cba"
> (aref "abc" -1)

#\NUL
> (aref "abc" 100)

#\s
> (svref "abc" 0)

#\a
> (concatenate 'string '(1 2 3))

Condition of type: SIMPLE-PROGRAM-ERROR
expected class_id 113 symbol was not defined

Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[118] TOP-LEVEL.
 File: "/opt/clasp/build/clasp/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (coerce "abc" 'bit-vector)

#<vector 0x7f2ad4404bb1>
> (coerce '(1 0 1) 'bit-vector)

#<vector 0x7f2ad4542b01>
> (sbit "abc" 0)

#\a
> (subseq #*101 0 1)

Condition of type: SIMPLE-PROGRAM-ERROR
File(include/clasp/core/lispVector.h) lineNumber(99): Subclass[COMMON-LISP:SIMPLE-BIT-VECTOR] must implement method[subseq] 

Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[116] TOP-LEVEL.
 File: "/opt/clasp/build/clasp/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (subseq #(a b c) 0 1)

#<vector 0x7f2ad43d03f1>
> (equalp #(a b c) #(a b c))

Condition of type: SIMPLE-PROGRAM-ERROR
Implement me!!! Implement Array_O::equalp
function(equalp) file(src/core/array.cc) lineNumber(270)

Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[116] TOP-LEVEL.
 File: "/opt/clasp/build/clasp/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (equal #*111 #*111)

T
> (equalp #*111 #*111)   

Condition of type: SIMPLE-PROGRAM-ERROR
Implement me!!! Implement Array_O::equalp
function(equalp) file(src/core/array.cc) lineNumber(270)

Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[116] TOP-LEVEL.
 File: "/opt/clasp/build/clasp/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (equalp 1 1.0)

NIL
> (eql -0.0 0.0)

T
> 
```

**Comment by drmeister** [comment]
(sbit "abc" 0) will need to be fixed later using Cleavir

**Comment by Shinmera** [comment]
Does that mean the rest of the issues are solved?
Why does it require cleavir? What's the catch?

**Comment by drmeister** [comment]
ECL uses this code for sbit:
(defun sbit (bit-array &rest indices)
  "Args: (simple-bit-array &rest subscripts)
Returns the specified bit in SIMPLE-BIT-ARRAY."
  (declare (array bit-array) ;; FIXME! Should be (simple-array bit)
           (ext:check-arguments-type))
  (row-major-aref bit-array (row-major-index-inner bit-array indices)))

The arguments are checked by the compiler inserting code to do the check.  I can't really do this in C++.

**Comment by Shinmera** [comment]
Some more fixed, some more still broken on 62a597141aa08d54ae16277cd83c191ad447c3b6
```
> (nreverse "abc")

"cba"
> (reverse "abc")

"cba"
> (aref "abc" -1)

Condition of type: SIMPLE-PROGRAM-ERROR
Index -1 out of bounds - must be [0,3)

Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[115] TOP-LEVEL.
 File: "/home/linus/.cache/shirakumo/autobuild/clasp-testing/.autobuild/62a597141aa08d54ae16277cd83c191ad447c3b6/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (svref "abc" 0)

Condition of type: SIMPLE-TYPE-ERROR
In svref ../../include/clasp/core/str.h line 189
COMMON-LISP:TYPE-ERROR :initializers (:DATUM "abc" :EXPECTED-TYPE COMMON-LISP:SIMPLE-VECTOR  )


Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[115] TOP-LEVEL.
 File: "/home/linus/.cache/shirakumo/autobuild/clasp-testing/.autobuild/62a597141aa08d54ae16277cd83c191ad447c3b6/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (concatenate 'string '(1 2 3))

Condition of type: SIMPLE-TYPE-ERROR
In lisp_errorUnexpectedType ../../src/core/foundation.cc line 161
COMMON-LISP:TYPE-ERROR :initializers (:DATUM 1 :EXPECTED-TYPE COMMON-LISP:CHARACTER  )


Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[117] TOP-LEVEL.
 File: "/home/linus/.cache/shirakumo/autobuild/clasp-testing/.autobuild/62a597141aa08d54ae16277cd83c191ad447c3b6/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (coerce "abc" 'bit-vector)

Condition of type: SIMPLE-TYPE-ERROR
In lisp_errorUnexpectedType ../../src/core/foundation.cc line 161
COMMON-LISP:TYPE-ERROR :initializers (:DATUM #\a :EXPECTED-TYPE COMMON-LISP:FIXNUM  )


Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[118] TOP-LEVEL.
 File: "/home/linus/.cache/shirakumo/autobuild/clasp-testing/.autobuild/62a597141aa08d54ae16277cd83c191ad447c3b6/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (sbit "abc" 0)

#\a
> (subseq #*101 0 1)

#<bit-vector 0x2ac4896e62c9>
> (subseq #(a b c) 0 1)

#<vector 0x2ac48af81429>
> (equalp #(a b c) #(a b c))

Condition of type: SIMPLE-PROGRAM-ERROR
Implement me!!! Implement Array_O::equalp
function(equalp) file(../../src/core/array.cc) lineNumber(271)

Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[115] TOP-LEVEL.
 File: "/home/linus/.cache/shirakumo/autobuild/clasp-testing/.autobuild/62a597141aa08d54ae16277cd83c191ad447c3b6/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (equal #*111 #*111)

T
> (equalp #*111 #*111)   

Condition of type: SIMPLE-PROGRAM-ERROR
Implement me!!! Implement Array_O::equalp
function(equalp) file(../../src/core/array.cc) lineNumber(271)

Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[115] TOP-LEVEL.
 File: "/home/linus/.cache/shirakumo/autobuild/clasp-testing/.autobuild/62a597141aa08d54ae16277cd83c191ad447c3b6/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (equalp 1 1.0)

NIL
> (eql -0.0 0.0)

NIL
> 
```
So, still broken: `(equalp 1 1.0) (equalp #*111 #*111) (equalp #(a b c) #(a b c)) (subseq #(a b c) 0 1) (subseq #*101 0 1)`

**Comment by Shinmera** [comment]
Two more down, three to go still on ca38835456a53665163cef96907c203cd7c50ef2
```
> (equalp 1 1.0)

T
> (equalp #*111 #*111)

Condition of type: SIMPLE-PROGRAM-ERROR
File(../../include/clasp/core/array.h) lineNumber(125): Subclass[COMMON-LISP:SIMPLE-BIT-VECTOR] must implement method[svref] 

Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[115] TOP-LEVEL.
 File: "/home/linus/.cache/shirakumo/autobuild/clasp-testing/.autobuild/ca38835456a53665163cef96907c203cd7c50ef2/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (equalp #(a b c) #(a b c))

T
> (subseq #(a b c) 0 1)

#<vector 0x2b9a32d894b9>
> (subseq #*101 0 1)

#<bit-vector 0x2b9a3234bcc9>
```

**Comment by drmeister** [comment]
I added svref to bit-vectors and all of the other problems in this issue should be fixed as of commit 0f27721735b30f92606b3fa23720d1704d91abaa

**Comment by Shinmera** [comment]
OH! So close! 0f27721735b30f92606b3fa23720d1704d91abaa
```
> (equalp 1 1.0)

T
> (equalp #*111 #*111)

T
> (equalp #(a b c) #(a b c))

T
> (subseq #(a b c) 0 1)

#(A)
> (subseq #*101 0 1)

#*1
> (nreverse "abc")

"cba"
> (reverse "abc")

"cba"
> (aref "abc" -1)

Condition of type: SIMPLE-PROGRAM-ERROR
Index -1 out of bounds - must be [0,3)

Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[120] TOP-LEVEL.
 File: "/home/linus/.cache/shirakumo/autobuild/clasp-testing/.autobuild/0f27721735b30f92606b3fa23720d1704d91abaa/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (aref "abc" 100)

Condition of type: SIMPLE-PROGRAM-ERROR
Index 100 out of bounds - must be [0,3)

Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[120] TOP-LEVEL.
 File: "/home/linus/.cache/shirakumo/autobuild/clasp-testing/.autobuild/0f27721735b30f92606b3fa23720d1704d91abaa/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (svref "abc" 0)

Condition of type: SIMPLE-TYPE-ERROR
In svref ../../include/clasp/core/str.h line 189
COMMON-LISP:TYPE-ERROR :initializers (:DATUM "abc" :EXPECTED-TYPE COMMON-LISP:SIMPLE-VECTOR  )


Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[120] TOP-LEVEL.
 File: "/home/linus/.cache/shirakumo/autobuild/clasp-testing/.autobuild/0f27721735b30f92606b3fa23720d1704d91abaa/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (concatenate 'string '(1 2 3))

Condition of type: SIMPLE-TYPE-ERROR
In lisp_errorUnexpectedType ../../src/core/foundation.cc line 161
COMMON-LISP:TYPE-ERROR :initializers (:DATUM 1 :EXPECTED-TYPE COMMON-LISP:CHARACTER  )


Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[122] TOP-LEVEL.
 File: "/home/linus/.cache/shirakumo/autobuild/clasp-testing/.autobuild/0f27721735b30f92606b3fa23720d1704d91abaa/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (coerce "abc" 'bit-vector)

Condition of type: SIMPLE-TYPE-ERROR
In lisp_errorUnexpectedType ../../src/core/foundation.cc line 161
COMMON-LISP:TYPE-ERROR :initializers (:DATUM #\a :EXPECTED-TYPE COMMON-LISP:FIXNUM  )


Available restarts:
(use :r1 to invoke restart 1)

1. (RESTART-TOPLEVEL) Go back to Top-Level REPL.

Broken at frame[123] TOP-LEVEL.
 File: "/home/linus/.cache/shirakumo/autobuild/clasp-testing/.autobuild/0f27721735b30f92606b3fa23720d1704d91abaa/build/clasp/Contents/Resources/lisp/kernel/lsp/top.lsp" (Position #21032)
>> :r1

> (coerce '(1 0 1) 'bit-vector)

#*101
> (sbit "abc" 0)

#\a
> (subseq #*101 0 1)

#*1
> (subseq #(a b c) 0 1)

#(A)
> (equalp #(a b c) #(a b c))

T
> (equal #*111 #*111)

T
> (equalp #*111 #*111)   

T
> (equalp 1 1.0)

T
> (eql -0.0 0.0)

NIL
> 
```
`(sbit "abc" 0)` should error, but doesn't.

**Comment by drmeister** [comment]
I fixed bit and sbit by adding a check-type macro on their argument.  The fix is in https://github.com/drmeister/clasp/commit/dea135513350b3509bc837edf4ca34bcf530158a

**Comment by Shinmera** [comment]
Fixed on dea135513350b3509bc837edf4ca34bcf530158a
```
> (sbit #*101 0)

1
> (sbit "abc" 0)

Condition of type: SIMPLE-TYPE-ERROR
The value of CORE::BIT-ARRAY is "abc", which is not of type (SIMPLE-ARRAY BIT).
```
